include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,382 @@
+// Copyright (C) 2020 - AnastaZIuk
+// This file is part of the "IrrlichtBAW" engine.
+// For conditions of distribution and use, see copyright notice in irrlicht.h
+
+#ifndef __IRR_C_SUMMED_AREA_TABLE_IMAGE_FILTER_H_INCLUDED__
+#define __IRR_C_SUMMED_AREA_TABLE_IMAGE_FILTER_H_INCLUDED__
+
+#include "irr/core/core.h"
+
+#include <type_traits>
+#include <functional>
+
+#include "irr/asset/filters/CMatchedSizeInOutImageFilterCommon.h"
+#include "CConvertFormatImageFilter.h"
+
+namespace irr
+{
+namespace asset
+{
+
+namespace impl
+{
+	struct plusTexels
+	{
+		core::vector4df_SIMD operator()(const core::vector4df_SIMD& lhs, const core::vector4df_SIMD& rhs) const	// it will fuck up 100% XD, TODO
+		{
+			return lhs + rhs;
+		}
+	};
+}
+
+template<bool ExclusiveMode>
+class CSummedAreaTableImageFilterBase : public CBasicImageFilterCommon
+{
+	public:
+		class CStateBase
+		{
+			public:
+				enum E_SUM_MODE
+				{
+					ESM_INCLUSIVE,	//!< all the values are summed withing the pixel summing begin from, so (x,y,z) values <= than itself
+					ESM_EXCLUSIVE,	//!< all the values are summed without the pixel summing begin from, so (x,y,z) values < than itself
+					EAS_COUNT
+				};
+
+				uint8_t*	scratchMemory = nullptr;										//!< memory used for temporary filling within computation of sum values
+				uint32_t	scratchMemoryByteSize = 0u;
+				core::smart_refctd_ptr<asset::ICPUBuffer>	rowSumCache;					//!< pointer of row extent size for y values used for holding max values going down till 0 is achievied
+				core::smart_refctd_ptr<asset::ICPUBuffer>	columnSumCache;					//!< pointer of column extent size for x values used for holding max values going down till 0 is achievied
+				constexpr E_SUM_MODE mode = ExclusiveMode ? ESM_EXCLUSIVE : ESM_INCLUSIVE;
+		};
+
+	protected:
+		CSummedAreaTableImageFilterBase() {}
+		virtual ~CSummedAreaTableImageFilterBase() {}
+
+		static inline uint32_t getRequiredScratchByteSize(E_FORMAT outFormat, const core::vectorSIMDu32& outExtent = core::vectorSIMDu32(0,0,0,0))
+		{
+			return asset::getTexelOrBlockBytesize(outFormat) * outExtent.X * outExtent.Y * outExtent.Z;
```

Comment:
its going to be `8u` always, you will do the summation in `double` or `uint64_t`

Reply:
corrected in following commit